packages/gatsby/src/datastore/in-memory/indexing.ts

```diff
@@ -21,6 +24,88 @@ export type FilterOp =  // TODO: merge with DbComparator ?
 // Note: `undefined` is an encoding for a property that does not exist
 
 export type FilterCacheKey = string
+type GatsbyNodeID = string
+
+export interface IGatsbyNodePartial {
+  isGatsbyNodePartial: boolean
+  id: GatsbyNodeID
+  internal: {
+    counter: number
+  }
+  indexFields: Set<string>
+  [k: string]: any
+}
+
+const nodeIdToIdentifierMap = new Map<
+  GatsbyNodeID,
+  WeakRef<IGatsbyNodePartial>
+>()
+
+/**
+ * Grabs an instance of IGatsbyNodePartial for the given node.
+ * This accepts an IGatsbyNode or IGatsbyNodePartial as input, which allows
+ * us to conditionally store index fields on the partial if we encounter
+ * one that hasn't been stored on the partial yet.
+ */
+export const getGatsbyNodePartial = (
+  node: IGatsbyNode | IGatsbyNodePartial,
+  indexFields: Array<string>,
+  resolvedFields: any
+): IGatsbyNodePartial => {
+  // first, check if we have the partial in the cache
+  const cacheKey = `${node.id}_____${node.internal.counter}`
+  if (nodeIdToIdentifierMap.has(cacheKey)) {
+    const maybeStillExist = nodeIdToIdentifierMap.get(cacheKey)?.deref()
+
+    // now check if we have it in memory and it has all the fields we need
+    if (
+      maybeStillExist &&
+      _.isEqual(new Set(indexFields), maybeStillExist.indexFields)
+    ) {
+      return maybeStillExist
+    }
```

Comment:
Question - if `indexFields` do not match, we will create new partial and overwrite `nodeIdToIdentifierMap` entry with it, losing previous one.

Consider scenario (not sure if this can happen) - order of events:
 - `nodeIdToIdentifierMap ` map is empty
 - We try to get partial with `['foo']` index fields - we populate map
 - We try to get partial with `['bar']` index fields - we create new partial and overwrite entry created by previous step
 - We try to get partial with `['foo']` index fields again - we create new partial again
 
 I'm not sure if this can happen (I would assume so). Should we either have cache key that takes `indexFields` into consideration? Or maybe we could have same cache key and instead "merge" one we have and just add new `indexFields` to it?

Reply:
Ahh you're right. We shouldn't do a strict equals, but instead check if the new `indexFields` is a subset of what we have already.